src/shared/utils/ast/visitors.ts

```diff
@@ -0,0 +1,388 @@
+import {isObject} from 'lodash'
+import {format_from_js_file} from '@influxdata/flux-lsp-browser'
+import {
+  Node,
+  CallExpression,
+  VariableAssignment,
+  ObjectExpression,
+  Expression,
+  Property,
+  Identifier,
+  StringLiteral,
+} from 'src/types/ast'
+import {WINDOW_PERIOD} from 'src/variables/constants'
+import {
+  isRangeNode,
+  constructWindowVarAssignmentFromRange,
+} from 'src/shared/utils/ast'
+
+/*
+  Legacy function.
+
+  Find all nodes in a tree matching the `predicate` function.
+    * Each node in the tree is an object.
+    * Each object which may contain objects or arrays of objects as children, under any key.
+*/
+export const findNodes = (
+  node: any,
+  predicate: (node: any) => boolean,
+  acc: any[] = []
+) => {
+  if (predicate(node)) {
+    acc.push(node)
+  }
+
+  for (const value of Object.values(node)) {
+    if (isObject(value)) {
+      findNodes(value as Node, predicate, acc)
+    } else if (Array.isArray(value)) {
+      for (const innerValue of value) {
+        findNodes(innerValue, predicate, acc)
+      }
+    }
+  }
+
+  return acc
+}
+
+export type ScopeVariableAssignedNodeT =
+  | VariableAssignment
+  | Expression
+  | Property
+
+export interface AstScope {
+  [variableIdentifier: string]: ScopeVariableAssignedNodeT
+}
+interface RequiredAcc {
+  halted: boolean
+  scope: AstScope
+}
+interface Accumulator extends RequiredAcc {
+  [callerProps: string]: any
+}
+
+const SCOPED_WINDOW_PERIOD = `v.${WINDOW_PERIOD}`
+
+/**
+ * Find the first node matching the halt `predicate` function.
+ * Return the variable scope for this node.
+ *
+ * @param node -- current node
+ * @param halt -- halt condition
+ * @param whenFound -- callback triggered when halt condition is met
+ * @param {Accumulator} acc -- accumulator
+ * @returns {Accumulator} -- return Accumulator, including scope
+ */
+export const findNodeScope = (
+  node: Node,
+  halt: (node: any) => boolean,
+  whenFound: (node, acc) => Accumulator,
+  acc: Accumulator = {halted: false, scope: {} as AstScope}
+): Accumulator => {
+  if (acc.halted) {
+    return acc
+  }
+  if (halt(node)) {
+    return whenFound(node, {...acc, halted: true})
+  }
+
+  // have seen !node.type with range subtrees, and lambda expr
+  if (!node.type) {
+    let fn = recurseWithoutSharingSiblingScope
+    if (isRangeNode(node)) {
+      if (!acc.scope[SCOPED_WINDOW_PERIOD]) {
+        const windowVar = constructWindowVarAssignmentFromRange(
+          acc.scope,
+          node as CallExpression
+        )
+        acc.scope = {...acc.scope, [SCOPED_WINDOW_PERIOD]: windowVar}
+        // TODO: fix tech debt. Breaking the interface of this visitor.
+        acc.toInjectWindowVar = true
+      }
+      fn = recurseWITHSharingSiblingScope
+    }
+    if (node.hasOwnProperty('arguments')) {
+      return fn((node as any).arguments, halt, whenFound, acc)
+    }
+    return acc
+  }
+
+  switch (node.type) {
```

Comment:
This is one of the classic approaches to walking an AST tree. AST is a tree of nodes, with each node having a type. Know how to handle each node type (independent of the rest of the tree), and then recursively call the immediate children of this node.